idea/idea-analysis/src/org/jetbrains/kotlin/idea/caches/KotlinShortNamesCache.kt

```diff
@@ -85,30 +87,78 @@ class KotlinShortNamesCache(private val project: Project) : PsiShortNamesCache()
-            = emptyArray()
+        val propertyAccessorsPsi = getPropertyNamesCandidatesByAccessorName(Name.identifier(name))
+                .flatMap { propertiesIndex.get(it.asString(), project, scope) }
+                .flatMap { LightClassUtil.getLightClassPropertyMethods(it).allDeclarations }
+                .filter { it.name == name }
+                .map { it as? PsiMethod }
+
+        return (propertyAccessorsPsi + kotlinFunctionsPsi).filterNotNull().toList().toTypedArray()
+    }
+
+    override fun getMethodsByNameIfNotMoreThan(name: String, scope: GlobalSearchScope, maxCount: Int): Array<PsiMethod> {
+        val methods = getMethodsByName(name, scope)
+        if (methods.size > maxCount)
+            return PsiMethod.EMPTY_ARRAY
+        return methods
+    }
+
+    override fun getFieldsByNameIfNotMoreThan(name: String, scope: GlobalSearchScope, maxCount: Int): Array<PsiField> {
+        val fields = getFieldsByName(name, scope)
+        if (fields.size > maxCount)
+            return PsiField.EMPTY_ARRAY
+        return fields
+    }
 
     override fun processMethodsWithName(name: String, scope: GlobalSearchScope, processor: Processor<PsiMethod>): Boolean
             = ContainerUtil.process(getMethodsByName(name, scope), processor)
 
-    override fun getAllMethodNames(): Array<String>
-            = emptyArray()
+    override fun getAllMethodNames(): Array<String> {
+        val functionIndex = KotlinFunctionShortNameIndex.getInstance()
+        val functionNames = functionIndex.getAllKeys(project)
+
+        val propertiesIndex = KotlinPropertyShortNameIndex.getInstance()
+        val globalScope = GlobalSearchScope.allScope(project)
+        val propertyAccessorNames = propertiesIndex.getAllKeys(project)
+                .flatMap { propertiesIndex.get(it, project, globalScope) }
+                .flatMap { it.getAccessorNames() }
+
+        return (functionNames + propertyAccessorNames).toTypedArray()
+    }
 
     override fun getAllMethodNames(set: HashSet<String>) {
+        set.addAll(allMethodNames)
+    }
+
+    override fun getFieldsByName(name: String, scope: GlobalSearchScope): Array<PsiField> {
+        val propertiesIndex = KotlinPropertyShortNameIndex.getInstance()
+
+        return propertiesIndex.get(name, project, scope)
+                .map { LightClassUtil.getLightClassBackingField(it) }
+                .filterNotNull()
+                .toTypedArray()
     }
 
-    override fun getFieldsByName(name: String, scope: GlobalSearchScope): Array<PsiField>
-            = emptyArray()
+    override fun getAllFieldNames(): Array<String> {
+        val propertiesIndex = KotlinPropertyShortNameIndex.getInstance()
+        val globalScope = GlobalSearchScope.allScope(project)
 
-    override fun getAllFieldNames(): Array<String>
-            = emptyArray()
+        return propertiesIndex.getAllKeys(project)
+                .flatMap { propertiesIndex.get(it, project, globalScope) }
```

Comment:
But we unable to determine is property has an @JvmField and it's backing field available for access without having instance of KtProperty.
Or we should return all possible names without this check?  Or some approach like `KotlinPropertyShortNameIndex.getInstance().getAllValues(project)` is possible?